src/progress.rs

```diff
@@ -195,13 +195,13 @@ impl ProgressSet {
 
     /// Promote a learner to a peer.
     pub fn promote_learner(&mut self, id: u64) -> Result<(), Error> {
-        if self.learner_ids().contains(&id) {
-            self.configuration.voters.insert(id);
-            self.configuration.learners.remove(&id);
-        } else if self.voter_ids().contains(&id) {
+        if !self.configuration.learners.remove(&id) {
+            // Wasn't already a voter. We can't promote what doesn't exist.
+            return Err(Error::Exists(id, "learners"));
+        }
+        if !self.configuration.voters.insert(id) {
```

Comment:
This should always be false, which is guaranteed by `assert_progress_and_configuration_consistent`.

Reply:
`assert_progress_and_configuration_consistent` validates conditions on the end of an operation.

This checks for a case such as:

```rust
prs.insert_voter(1); // Whoops! I meant learner.
prs.promote_learner(1); // -> Gives an Error::Exists(1, "voters")
```

Reply:
The `assertion` appears everywhere, so the correctness should always be hold. The example you provide will return error at L200.

Reply:
The error at `L200` appears from:

```rust
// Do **not** add 1 as a learner.
prs.promote_learner(1); // Err(Error::Exists(id, "learners"))
```

The conditions of this function are:

* The `id` must currently be a learner. (Checked on L198),
* The `id` is not already a voter. (Checked here on L202),

Reply:
Can you write a test case that will make this function return `Err(Error::Exists(id, "voters"))`? From what I can see, this is impossible via the public APIs.

Reply:
Using `Raft::add_node()` or `Raft::add_voter()` which call `Raft::add_voter_or_learner()` cannot result in this error, it's true.

Note that `raft.add_node(1);  raft.prs().promote_learner(1);` is completely possible for a user to do. So we must protect from it.